src/shims/posix/linux/foreign_items.rs

```diff
@@ -208,7 +208,11 @@ fn getrandom<'tcx>(
 
     // The only supported flags are GRND_RANDOM and GRND_NONBLOCK,
     // neither of which have any effect on our current PRNG.
-    this.read_scalar(flags)?.to_i32()?;
+    let flags = this.read_scalar(flags)?;
+    // Either `i32` or `isize` is fine.
+    if flags.to_machine_isize(this).is_err() {
```

Comment:
This is extremely suspicious, I will have to think about what to do here. `is_err` is never okay; Miri should not catch exceptions (that leads to all sots of problems).

Reply:
yea, I think the right fix is to check `flags.ty.kind()` and then choose a different `to_*` method

Reply:
First I need to become [convinced that this is not UB](https://github.com/rust-lang/rust/pull/78785#pullrequestreview-534176066).

Reply:
Variadic functions like this, especially `syscall`, are pretty annoying. For the ABI it doesn't matter if you pass a `char` or `int` or `long` or a pointer, those just all count as 'one argument' or 'one register'. But unless miri is going to emulate the actual processor registers to simulate the ABI, this is always going to be a bit hacky. :( (The futex implementation had the problem with accepting an integer for the pointer argument.)

Reply:
I don't know if the C standard defines this behaviour, but the `syscall` wrapper isn't a C function. It's written in assembly and fully depends on knowing everything about the ABI.

Reply:
> (The futex implementation had the problem with accepting an integer for the pointer argument.)

Those are the same size, so they are the same ABI even as far as Rust's Layout/ABI code is concerned. Thus, no problem here.

This is different. I spent a lot of time making sure Miri detects ABI mismatches during function calls, and I'd consider different sizes to be ABI mismatches. You can't even transmute these!

Reply:
> I don't know if the C standard defines this behaviour, but the syscall wrapper isn't a C function. It's written in assembly and fully depends on knowing everything about the ABI.

It is called with the `"C"` ABI though, so its rules apply.

Reply:
Also note that Miri's `getrandom` function is also used to shim the `getrandom` libc function, not just direct calls to the syscall.

Reply:
> > I don't know if the C standard defines this behaviour, but the syscall wrapper isn't a C function. It's written in assembly and fully depends on knowing everything about the ABI.
> 
> It is called with the `"C"` ABI though, so its rules apply.

@RalfJung From what little I know of the x86_64 `"C"` ABI, you can call an `extern "C" fn(u64)` as if it was an `extern "C" fn(u8)` because in both cases the argument is passed in the same 64-bit register[0]. I think that is what @m-ou-se meant by "fully depends on knowing everything about the ABI."

[0]: See e.g. https://godbolt.org/z/a35MYb, keeping in mind that `edi` (resp. `eax`) is an alias to the low 32 bits of `rdi` (resp. `rax`).

Reply:
> you can call an extern "C" fn(u64) as if it was an extern "C" fn(u8) because in both cases the argument is passed in the same 64-bit register[0]

And which 64-bit value will the function see then? Maybe the compiler passes some junk in the higher bytes?

Also, the code we are looking at in rustc runs on many target. So just looking at the x86_64 ABI does not help.